File: tests/pps-conformance/validate.js
// Minimal JSON Schema validation for PPS envelopes
// Usage: node tests/pps-conformance/validate.js <path-to-envelope.json>

const fs = require('fs');
const path = require('path');
const Ajv = require('ajv/dist/2020');

function loadJson(filePath) {
  const abs = path.resolve(process.cwd(), filePath);
  return JSON.parse(fs.readFileSync(abs, 'utf8'));
}

function main() {
  const target = process.argv[2];
  if (!target) {
    console.error('Usage: validate.js <path-to-envelope.json>');
    process.exit(2);
  }

  const schema = loadJson('spec/pps.schema.json');
  const data = loadJson(target);

  const ajv = new Ajv({ allErrors: true, strict: false, validateFormats: false });
  const validate = ajv.compile(schema);
  const ok = validate(data);

  if (ok) {
    console.log('VALID');
  } else {
    console.log('INVALID');
    console.log(JSON.stringify(validate.errors, null, 2));
    process.exit(1);
  }
}

main();

File: sdk/python/pps/_validate.py
"""JSON Schema validation for PPS envelopes."""

import jsonschema
from ._schema import PPS_SCHEMA


def validate(envelope: dict) -> dict:
    """
    Validate a PPS envelope against the JSON Schema.

    Returns:
        {"valid": True, "errors": []}
        {"valid": False, "errors": [{"path": ..., "message": ...}, ...]}
    """
    validator = jsonschema.Draft202012Validator(PPS_SCHEMA)
    errors = list(validator.iter_errors(envelope))
    if not errors:
        return {"valid": True, "errors": []}
    return {
        "valid": False,
        "errors": [
            {
                "path": "/" + "/".join(str(p) for p in e.absolute_path) if e.absolute_path else "/",
                "message": e.message,
                "schema_path": list(e.absolute_schema_path),
            }
            for e in errors
        ],
    }
